File: main/src/main/java/com/sedmelluq/discord/lavaplayer/format/transcoder/OpusChunkDecoder.java
package com.sedmelluq.discord.lavaplayer.format.transcoder;

import com.sedmelluq.discord.lavaplayer.format.AudioDataFormat;
import com.sedmelluq.discord.lavaplayer.natives.opus.OpusDecoder;

import java.nio.ByteBuffer;
import java.nio.ShortBuffer;

/**
 * Audio chunk decoder for Opus codec.
 */
public class OpusChunkDecoder implements AudioChunkDecoder {
  private final OpusDecoder decoder;
  private final ByteBuffer encodedBuffer;

  /**
   * @param format Source audio format.
   */
  public OpusChunkDecoder(AudioDataFormat format) {
    encodedBuffer = ByteBuffer.allocateDirect(4096);
    decoder = new OpusDecoder(format.sampleRate, format.channelCount);
  }

  @Override
  public void decode(byte[] encoded, ShortBuffer buffer) {
    encodedBuffer.clear();
    encodedBuffer.put(encoded);
    encodedBuffer.flip();

    buffer.clear();
    decoder.decode(encodedBuffer, buffer);
  }

  @Override
  public void close() {
    decoder.close();
  }
}

File: main/src/main/java/lavaplayer/format/transcoder/OpusChunkDecoder.kt
package lavaplayer.format.transcoder

import lavaplayer.format.AudioDataFormat
import lavaplayer.natives.opus.OpusDecoder
import java.nio.ByteBuffer
import java.nio.ShortBuffer

/**
 * Audio chunk decoder for Opus codec.
 *
 * @param format Source audio format.
 */
class OpusChunkDecoder(format: AudioDataFormat) : AudioChunkDecoder {
    private val decoder = OpusDecoder(format.sampleRate, format.channelCount)
    private val encodedBuffer: ByteBuffer = ByteBuffer.allocateDirect(4096)

    override fun decode(encoded: ByteArray, output: ShortBuffer) {
        encodedBuffer.clear()
        encodedBuffer.put(encoded)
        encodedBuffer.flip()
        output.clear()
        decoder.decode(encodedBuffer, output)
    }

    override fun close() {
        decoder.close()
    }
}
